libeos-updater-util/flatpak.c

```diff
@@ -1457,6 +1457,651 @@ euu_filter_for_existing_flatpak_ref_actions (GHashTable *ref_actions  /* (elemen
+                                            flatpak_ref_get_arch (ref),
+                                            flatpak_ref_get_branch (ref),
+                                            cancellable,
+                                            &local_error);
+
+  if (installed_ref == NULL)
+   {
+     if (g_error_matches (local_error, FLATPAK_ERROR, FLATPAK_ERROR_NOT_INSTALLED))
+       {
+         g_clear_error (&local_error);
+         return TRUE;
+       }
+
+      g_propagate_error (error, g_steal_pointer (&local_error));
+      return FALSE;
+   }
+
+  for (i = 0; i < remotes->len; ++i)
+    {
+      FlatpakRemote *remote = g_ptr_array_index (remotes, i);
+      g_autoptr(GPtrArray) installed_related_refs =
+        list_installed_related_refs_for_remote (installation,
+                                                flatpak_remote_get_name (remote),
+                                                ref,
+                                                formatted_ref,
+                                                cancellable,
+                                                error);
+      guint j = 0;
+
+      if (installed_related_refs == NULL)
+        return TRUE;
+
+      for (; j < installed_related_refs->len; ++j)
+        {
+          FlatpakRef *related_ref = FLATPAK_REF (g_ptr_array_index (installed_related_refs, j));
+
+          /* We can avoid a rather expensive recursive call here
+           * if we check now, so check before continuing */
+          if (g_hash_table_contains (installed_related_refs_set, related_ref))
+            continue;
+
+          g_hash_table_insert (installed_related_refs_set,
+                               g_object_ref (related_ref),
+                               NULL);
+
+          /* Populate installed related refs for this related ref */
+          if (!populate_installed_related_refs (installation,
+                                                related_ref,
+                                                remotes,
+                                                installed_related_refs_set,
+                                                cancellable,
+                                                error))
+            return FALSE;
+        }
+    }
+
+  return TRUE;
+}
+
+static gboolean
+determine_action_for_related_ref (FlatpakRelatedRef             *related_ref,
+                                  EuuFlatpakRemoteRefAction     *source_ref_action,
+                                  GHashTable                    *installed_related_refs,
+                                  EuuFlatpakRemoteRefActionType *out_action_type)
+{
+  g_return_val_if_fail (out_action_type != NULL, FALSE);
+
+  switch (source_ref_action->type)
+    {
+      case EUU_FLATPAK_REMOTE_REF_ACTION_INSTALL:
+        if (flatpak_related_ref_should_download (related_ref))
+          {
+            if (g_hash_table_contains (installed_related_refs, related_ref))
```

Comment:
You could factor part of this out above:
```
gboolean related_ref_is_installed = g_hash_table_contains (…);
```
and then use `related_ref_is_installed` in the `if` conditions.